File: Assets/Code/Logic/UI/Tips/View/CommonRewardTipsView.cs
using UnityEngine;
using System.Collections;
using UnityEngine.UI;
using Logic.UI.CommonReward.View;
using System.Collections.Generic;
using Logic.Enums;
using Logic.Game.Model;
using Common.Util;
using Common.Localization;

namespace  Logic.UI.Tips.View
{
	public class CommonRewardTipsView : MonoBehaviour 
	{
		public const string PREFAB_PATH = "ui/tips/common_reward_tips_view";
		
		public static CommonRewardTipsView Open(List<GameResData> dataList,bool rewardEnable = false, System.Action clickRewardCallback = null)
		{
			CommonRewardTipsView view = UIMgr.instance.Open<CommonRewardTipsView>(PREFAB_PATH, EUISortingLayer.Tips);
			view._clickRewardCallback = clickRewardCallback;
			view.SetRewardButtonEnable(rewardEnable);
			view.SetDataList(dataList);
			return view;
		}

		#region ui component
		public GameObject rewardIconPrefab;
		public Transform rewardIconRoot;
		public Button getRewardBtn;
		#endregion

		private System.Action _clickRewardCallback;

		void Start()
		{
//			List<GameResData> rewardList = new List<GameResData>();
//			rewardList.Add(new GameResData(Logic.Enums.BaseResType.Account_Exp,0,100,0));
//			rewardList.Add(new GameResData(Logic.Enums.BaseResType.Account_Exp,0,100,0));
//			rewardList.Add(new GameResData(Logic.Enums.BaseResType.Account_Exp,0,100,0));
//			rewardList.Add(new GameResData(Logic.Enums.BaseResType.Account_Exp,0,100,0));
//			rewardList.Add(new GameResData(Logic.Enums.BaseResType.Account_Exp,0,100,0));
//			rewardList.Add(new GameResData(Logic.Enums.BaseResType.Account_Exp,0,100,0));
//			rewardList.Add(new GameResData(Logic.Enums.BaseResType.Account_Exp,0,100,0));
//			SetDataList(rewardList);
		}

		private void Init()
		{


		}
		public void SetDataList(List<GameResData> dataList)
		{
			if(dataList == null)
			{
				return;
			}
			TransformUtil.ClearChildren(rewardIconRoot,true);
			int count = dataList.Count;
			int spaceHorizon = 130;
			int spaceVertical = 120;
			int columnMax = 4;
			int totalRow = count/columnMax + (count%columnMax == 0 ? 0 : 1) ;
			//
			GridLayoutGroup layout = rewardIconRoot.GetComponent<GridLayoutGroup>();

			float height = layout.cellSize.y+layout.spacing.y;
			RectTransform rectTran = rewardIconRoot.transform as RectTransform;
			Vector2 rewardRootSizeDelta = rectTran.sizeDelta;
			rewardRootSizeDelta = rectTran.sizeDelta = new Vector2(rewardRootSizeDelta.x, height*totalRow-layout.spacing.y+50);
				
			if(totalRow == 1)
			{
				if(layout!=null)
					layout.enabled = false;
			}else
			{
				if(layout!=null)
					layout.enabled = true;
			}
			rewardIconPrefab.SetActive(true);

			float xPadding = rewardRootSizeDelta.x/2;
			float yPadding = -100;
			for(int i = 0;i<count;i++)
			{
				GameObject clone = Instantiate<GameObject>(rewardIconPrefab);

				CommonRewardIcon icon = CommonRewardIcon.Create(clone.transform);
				RectTransform iconTran = icon.transform as RectTransform;
				iconTran.SetAsFirstSibling();
//				iconTran.anchorMax = iconTran.anchorMin = new Vector2(0.5f,0.5f);
				GameResData resData = dataList[i];
				icon.SetGameResData(resData);
				Transform tran = clone.transform;
				tran.SetParent(rewardIconRoot,false);
				float x = 0;
				float y = 0;

				int row = i/columnMax;
				int col = i%columnMax;
				if(totalRow == 1)//单行时居中
				{
					x = count%2 == 0 ?((i-count/2)*spaceHorizon+spaceHorizon/2):((i-count/2)*spaceHorizon);
				}else{//左对齐
					x = columnMax%2 == 0 ? ((col-columnMax/2)*spaceHorizon+spaceHorizon/2):((col-columnMax/2)*spaceHorizon);
					y = totalRow%2 == 0 ? ((totalRow/2-row)*spaceVertical-spaceVertical/2):((totalRow/2-row)*spaceVertical);
				}

				tran.localPosition = new Vector3(x+xPadding,y+yPadding);
				Text countText = tran.Find("text_count").GetComponent<Text>();
				countText.text = string.Format(Localization.Get("common.x_count"),resData.count);

				icon.ShowRewardCount(true);
			}
			rewardIconPrefab.SetActive(false);
		}
		private void SetRewardButtonEnable(bool enable)
		{
			getRewardBtn.enabled = enable;
			getRewardBtn.GetComponent<Image>().SetGray(!enable);
			Text text = getRewardBtn.GetComponentInChildren<Text>();
			if(text!= null)
			{
				text.color = enable ? Color.white : Color.gray;
				 Outline outline = text.GetComponent<Outline>();
				if(outline!=null)
					outline.enabled = false;
			}
		}
		public void OnClickCloseBtnHandler()
		{
			UIMgr.instance.Close(PREFAB_PATH);
		}
		public void OnClickGetRewardBtnHandler()
		{
			if(_clickRewardCallback!= null)
			{
				_clickRewardCallback();
			}

			UIMgr.instance.Close(PREFAB_PATH);
		}

	}

}

File: Assets/ToLua/Lua/user/ui/tips/view/common_reward_tips_view.lua
local t = {}
local PREFAB_PATH = 'ui/tips/common_reward_tips_view_lua'
t.__index=t

local game_res_data=require('ui/game/model/game_res_data')

--四段式列表(gameResData)
function t.Create(dataList)

  t.transform=Logic.UI.UIMgr.instance:Open(PREFAB_PATH,EUISortingLayer.Tips,UIOpenMode.Overlay).transform
  t.rewardIconPrefab=t.transform:FindChild('core/reward_templete').gameObject
  t.rewardIconRoot=t.transform:FindChild('core/Scroll View/content')
  t.transform:FindChild('core/title/text_name'):GetComponent(typeof(Text)).text = LuaInterface.LuaCsTransfer.LocalizationGet('event_public_award')
  t.btnClose = t.transform:FindChild('bg_core/btn_background'):GetComponent(typeof(Button))
  t.btnClose.onClick:AddListener(t.OnClickCloseHandler)
  
  coroutine.start(t.ShowBtnCoroutine)
  
  t.SetDataList(dataList)
  return t
end

--四段式列表(字符串)
function t.CreateOriginal(list)
  
  local datas = {}
  for i=1,#list do
    table.insert(datas,game_res_data.NewByString(list[i]))
  end
  return t.Create(datas)
end
--四段式列表(c# gameResData)
function t.CreateByCSharpGameResDataList(cSharpResDataList)
  local count = cSharpResDataList.Count
  local datas = {}
  for i = 0,count-1 do
    local resData = cSharpResDataList:get_Item(i)
    table.insert(datas,game_res_data.New(resData.type:ToInt(),resData.id,resData.count,resData.star))
  end
  return t.Create(datas)
end

function t.SetDataList(dataList)
  if not dataList then return end
  ui_util.ClearChildren(t.rewardIconRoot,true)
  local count = #dataList
  
  for  i=1,count do
    local clone = Instantiate(t.rewardIconPrefab)
    local resData = dataList[i]

    local icon = require('ui/common_icon/common_reward_icon').New(clone.transform,resData)
    icon:AddDesButton()

    local iconTran = icon.transform:GetComponent(typeof(RectTransform))
    iconTran:SetAsFirstSibling()
    local tran = clone.transform
    tran:SetParent(t.rewardIconRoot,false)
    tran.gameObject:SetActive(true)
    
    local countText = tran:FindChild("text_count"):GetComponent(typeof(Text))
    local nameText = tran:FindChild("text_name"):GetComponent(typeof(Text))
    countText.text = string.format(LuaInterface.LuaCsTransfer.LocalizationGet("common.x_count_lua"),resData.count)
    icon:HideCount() 
    
    local bData = nil    
    if resData.type == BaseResType.Hero then bData = gamemanager.GetData('hero_data').GetDataById(resData.id)
    elseif resData.type == BaseResType.Equipment then bData = gamemanager.GetData('equip_data').GetDataById(resData.id)
    elseif resData.type == BaseResType.Item then bData = gamemanager.GetData('item_data').GetDataById(resData.id) 
    else bData = gamemanager.GetData('item_data').GetBasicResItemByType(resData.type) end
    
    if nil ~= bData then nameText.text = LocalizationController.instance:Get(bData.name) end
  end

  t.rewardIconPrefab:SetActive(false)  
end

function t.SetRewardButtonEnable(enable)
  t.getRewardBtn.enabled = enable
  ui_util.SetImageGray(t.getRewardBtn:GetComponent(typeof(Image)),not enable)
  local text = t.getRewardBtn:GetComponentInChildren(typeof(Text))
  if text then
    if enable then
      text.color = Color.white
    else
      text.color = Color.gray
    end
    local outline = text:GetComponent(typeof(Outline))
    if outline then
      outline.enabled = false
    end
  end
end

function t.ShowBtnCoroutine()
  t.btnClose.enabled = false
  coroutine.wait(0.2)
  t.btnClose.enabled = true
end

function t.OnClickCloseHandler()
  Logic.UI.UIMgr.instance:Close(PREFAB_PATH)
end

return t